packages/sanity/src/core/validation/validateDocument.ts

```diff
@@ -49,71 +50,117 @@ export function resolveTypeForArrayItem(
 
 /**
- * Validates a document against the given schema
- *
- * @param getClient - Function that returns a Sanity client, based on a passed configuration
- * @param doc - Document to validate
- * @param schema - Schema to validate against
- * @param context - Context to use for validation
- * @returns A promise that resolves to an array of validation markers
  * @beta
  */
-export function validateDocument(
-  getClient: ValidateItemOptions['getClient'],
-  doc: SanityDocument,
-  schema: Schema,
-  context?: Pick<ValidationContext, 'getDocumentExists' | 'i18n'>,
-): Promise<ValidationMarker[]> {
-  return lastValueFrom(validateDocumentObservable(getClient, doc, schema, context))
+export interface ValidateDocumentOptions {
+  /**
+   * The document to be validated
+   */
+  document: SanityDocument
+  /**
+   * The workspace instance this document belongs to. The document will
+   * be validated against the schema in the given workspace.
+   */
+  workspace: Workspace
+  /**
+   * The factory function used to get a sanity client used in custom validators.
+   * If not provided, the one from the workspace will be used.
+   */
+  getClient?: (clientOptions: SourceClientOptions) => SanityClient
+  /**
+   * The function used to check to see if a reference is published before. If
+   * you're validating many documents in bulk, you may want to query for all
+   * document IDs first and provide your own implementation using those.
+   *
+   * If no function is provided a default one will be provided that utilizes
+   * the `getClient` function provided.
+   */
+  getDocumentExists?: (options: {id: string}) => Promise<boolean>
 }
 
 /**
- * Validates a document against the given schema, returning an Observable
+ * Validates a document against the given workspace. Returns an array of
+ * validation markers with a path, message, and validation level.
  *
- * @param getClient - Function that returns a Sanity client, based on a passed configuration
- * @param doc - Document to validate
- * @param schema - Schema to validate against
- * @param context - Context to use for validation
- * @returns An Observable that produces an array of validation markers
  * @beta
  */
-export function validateDocumentObservable(
-  getClient: ValidateItemOptions['getClient'],
-  doc: SanityDocument,
-  schema: Schema,
-  context?: Pick<ValidationContext, 'getDocumentExists' | 'i18n'>,
-): Observable<ValidationMarker[]> {
-  const documentType = schema.get(doc._type)
+export function validateDocument({
+  document,
+  workspace,
+  ...options
+}: ValidateDocumentOptions): Promise<ValidationMarker[]> {
+  // TODO: consider decorating this with a `client.fetch` concurrency limiter
+  const getClient = options.getClient || workspace.getClient
+
+  const defaultGetDocumentExists: ValidateDocumentOptions['getDocumentExists'] = ({id}) => {
+    const client = getClient({apiVersion: 'v2021-03-25'})
+    return client.fetch(`count(*[_id == $id]) > 0`, {id})
+  }
```

Comment:
Ah great call out. Yeah I'll definitely have to do another pass.

> I also see the todo two lines up, but think we may want to do implement some batching here right away - if this is called with the default getDocumentExists on a document with 300 refs, I think it will cause actually 300 requests?

Yeah that's the currently the case right now but we could re-use the [concurrency limiter](https://github.com/sanity-io/sanity/blob/02879fc9a8642ceaeaecd0ac0640fe296eb3a8a0/packages/sanity/src/_internal/cli/util/wrapWithConcurrencyLimiter.ts) I included in the follow up PR. I'll probably re-introduce it in this PR and rebase the other one.

> The whole observePaths implementation in the linked code may be overkill/have unintended side effects, but the chunking of document IDs and using the doc endpoint with excludeContent and checking the availability reason would be great to use.

Oh yeah that's perfect. I'll write something that batches a request there from separate calls. Thanks for the recommendation!